### PR TITLE
8274983: C1 optimizes the invocation of private interface methods

### PR DESCRIPTION
**Pull request by navyxliu**
The root cause of the C1 regression is that some regex generate multiple classes which all implement
an interface. In SlowStartupTest.java, the following **invokeinterface** happens frequently with different receivers. the target is a private interface method. 

```
9: invokeinterface #25,  3           // InterfaceMethod java/util/regex/Pattern$BmpCharPredicate.lambda$union$2:(Ljava/util/regex/Pattern$CharPredicate;I)Z
```

This patch allows c1 to generate the optimized virtual call for invokeinterface
whose targets are the private interface methods.

Before JDK-823835, LambdaMetaFactory generates invokespecial in this case. Because the private
interface methods can not be overrided, c1 generates the optimized virtual call. After JDK-823835,
LambdaMetaFactory generates invokeinterface instead. C1 generates the regular virtual call because
it can not recognize the new pattern. If a multiple of subclasses all implement a same interface,
it is possible that they trash the IC stub using their own concrete klass in runtime.

Optimized virtual call uses relocInfo::opt_virtual_call_type(3), It will call VM
'resolve_opt_virtual_call_C' once and resolve the target to the VEP of the nmethod.
Therefore, this patch can prevent the callsite from trashing.

Before this patch,   SlowStartupTest had 38770 times  _resolve_invoke_virtual_cnt and 38695  _handle_wrong_method_cnt per 10k iterations. To dump `C1Statistics`, we use fastdebug build for comparison.

```
$java -XX:TieredStopAtLevel=1 -XX:+PrintC1Statistics SlowStartupTest 1
Executed 10000 iterations in 736ms
C1 Runtime statistics:
 _resolve_invoke_virtual_cnt:     38770
 _resolve_invoke_opt_virtual_cnt: 186
 _resolve_invoke_static_cnt:      44
 _handle_wrong_method_cnt:        38695
 _ic_miss_cnt:                    35
```

With this patch,  only 1 _handle_wrong_method_cnt is triggered but we have 3 more `_resolve_invoke_opt_virtual_cnt` events instead.  The total runtime reduces from 736ms to 9ms. 

```
$java -XX:TieredStopAtLevel=1 -XX:+PrintC1Statistics SlowStartupTest 1
Executed 10000 iterations in 9ms
C1 Runtime statistics:
 _resolve_invoke_virtual_cnt:     77
 _resolve_invoke_opt_virtual_cnt: 189
 _resolve_invoke_static_cnt:      45
 _handle_wrong_method_cnt:        1
 _ic_miss_cnt:                    39
```

Codegen wise,  before the patch, C1 generates LIR for the invokeinterface whose target is a private interface methoda as follows.
```
__bci__use__tid____instr____________________________________
. 1    0    v2     a1.invokeinterface()
                   InvokePrivateInterfaceMethod$I.bar()V
. 6    0    v3     return
```

With this patch, C1 generates LIR as follows. it first check a1 is a subtype of `InvokePrivateInterfaceMethod$I`. if so, an optimized virtual call is generated. The callsite will be fixed up once and only one time in runtime. 
```
__bci__use__tid____instr____________________________________
. 1    1    a2     checkcast(a1) InvokePrivateInterfaceMethod$I
                   stack [0:a1]
. 1    0    v3     a2.invokeinterface()
                   InvokePrivateInterfaceMethod$I.bar()V
. 6    0    v4     return
```

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8274983](https://bugs.openjdk.java.net/browse/JDK-8274983): C1 optimizes the invocation of private interface methods

### Reviewers
 * [Dean Long](https://openjdk.java.net/census#dlong) (@dean-long - **Reviewer**) ⚠️ Review applies to 6a10e772862799edc1cfdc5c4455ffc4dbf1c6ed
 * [Igor Veresov](https://openjdk.java.net/census#iveresov) (@veresov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6445/head:pull/6445` \
`$ git checkout pull/6445`

Update a local copy of the PR: \
`$ git checkout pull/6445` \
`$ git pull https://git.openjdk.java.net/jdk pull/6445/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6445`

View PR using the GUI difftool: \
`$ git pr show -t 6445`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6445.diff">https://git.openjdk.java.net/jdk/pull/6445.diff</a>

</details>

### CODE REVIEW

**Comment by dean-long** [comment]
It looks like this does what we want for private interface methods, but I'm wondering if we handle all combinations of private/final and invokevirtual/invokeinterface, or are we missing some cases where can_be_statically_bound() would return true?

**Comment by navyxliu** [comment]
hi, @dean-long,

I think C1 covers all cases as long as the target method is loaded. I have seen cases which target methods haven't been loaded in startup time, but they are rare. 

ciMethod::can_be_statically_bound() return true if the method is private or final. The matrix shows the modifiers of target methods.

|                 | final | private | 
|-----------------|-------|---------|
| invokevirtual   | 1     | 2       | 
| invokespecial   | N/A1    | 3       |
| invokeinterface | N/A2    | 4       | 

1. generates the optimized virtual call because [x->target_is_final()](https://github.com/openjdk/jdk/blob/master/src/hotspot/share/c1/c1_LIRGenerator.cpp#L2799) is true.
2. transforms to `invokespecial` [here](https://github.com/openjdk/jdk/blob/master/src/hotspot/share/c1/c1_GraphBuilder.cpp#L1885) ,  then it will be case 3. 
3. generates the optimize virtual call because `x->code() == Bytecodes::_invokespecial` is true.
4. is what this patch covers. 

NA-1. I think it's impossible for javac.  it would be an optimized virtual call like case 3 even it existed.
NA-2: it's an illegal modifier for an interface method.
 https://docs.oracle.com/javase/specs/jls/se17/html/jls-9.html#jls-InterfaceMethodModifier

**Comment by dean-long** [comment]
Thanks @navyxliu.  I wonder if we can do 2) and 3) for invokeinterface, simplying the patch.  Something like:

  // Some methods are obviously bindable without any type checks so
  // convert them directly to an invokespecial or invokestatic.
  if (target->is_loaded() && !target->is_abstract() && target->can_be_statically_bound()) {
    switch (bc_raw) {
    case Bytecodes::_invokevirtual:
    case Bytecodes::_invokeinterface::
      code = Bytecodes::_invokespecial;
      break;
[...]
  // invoke-special-super
  if (code == Bytecodes::_invokespecial && !target->is_object_initializer()) {
    ciInstanceKlass* sender_klass = calling_klass;
    if (sender_klass->is_interface()) {
[...]

What do you think?

**Comment by navyxliu** [comment]
hi, @dean-long , 
I update the PR.  The new revision generate Invoke HIR with code == Invokespecial. I don't need to touch the LIR part, so I revert them. 

I call `c->set_incompatible_class_change_check()` instead of set_invokespecial_receiver_check(). this will call a runtime to throw ICCError directly. Before that,  it went to uncommon_trap and reinterpreted the bytecode.

File: src/hotspot/share/c1/c1_GraphBuilder.cpp [hunk 1]
```diff
@@ -1890,6 +1893,15 @@ void GraphBuilder::invoke(Bytecodes::Code code) {
     default:
       break;
     }
+
+    // Additional receiver subtype checks for interface calls via invokespecial or invokeinterface.
+    if (receiver_constraint != nullptr) {
+      int index = state()->stack_size() - (target->arg_size_no_receiver() + 1);
+      Value receiver = state()->stack_at(index);
+      CheckCast* c = new CheckCast(receiver_constraint, receiver, copy_state_before());
+      c->set_incompatible_class_change_check();
```
Comment (dean-long) [inline src/hotspot/share/c1/c1_GraphBuilder.cpp:1902]: set_incompatible_class_change_check() seems OK for invokeinterface, but for invokespecial, I believe the interpreter throws IllegalAccessError.

**Review by dean-long — CHANGES_REQUESTED** [review]

File: src/hotspot/share/c1/c1_GraphBuilder.cpp (hunk 1 above)
Comment (navyxliu) [reply src/hotspot/share/c1/c1_GraphBuilder.cpp]: Thanks to point it out.  

I take a look at invokespecial and invokeinterface part of JVM spec. There are multiple cases. I think I better use uncommon trap and let interpreter to cover them all. It's not a hot path anyway. 
 
By reverting this code, `java/lang/invoke/SpecialInterfaceCall.java` is fixed.

**Review by dean-long — APPROVED** [review]
As far as I can tell, this now matches what C2 does.

**Comment by navyxliu** [comment]
I add a JMH benchmark for this one.

```
$make test TEST='micro:org.openjdk.bench.vm.compiler.InterfacePrivateCalls' CONF=linux-x86_64-server-release

//master(7b2d823e)
Benchmark                                             Mode  Cnt   Score   Error  Units
InterfacePrivateCalls.invokePrivateInterfaceMethodC1  avgt    5  11399.058 ± 1086.745  ns/op
InterfacePrivateCalls.invokePrivateInterfaceMethodC2  avgt    5  23.741 ± 0.189  ns/op

// with this patch
Benchmark                                             Mode  Cnt   Score   Error  Units
InterfacePrivateCalls.invokePrivateInterfaceMethodC1  avgt    5  24.534 ± 0.130  ns/op
InterfacePrivateCalls.invokePrivateInterfaceMethodC2  avgt    5  23.800 ± 0.384  ns/op
```
The average cost of C1 has improved from 11399 ns/op to 24.534 ns/op, or 464x faster.  now the cost of c1 is comparable to C2.

**Review by veresov — APPROVED** [review]
Seems reasonable.

**Pull request closed without merging**